docs/manpages/tracker-index.1
.TH tracker-index 1 "September 2014" GNU "User Commands"

.SH NAME
tracker-index \- Index content using the Tracker filesystem miner

.SH SYNOPSIS
.nf
\fBtracker index\fR <\fIPATH\fR> [[\fIPATH\fR] ...]
\fBtracker index\fR [\-\-monitor] <\fIDIR\fR> [[\fIDIR\fR] ...]
.fi

.SH DESCRIPTION
This command causes the Tracker filesystem miner to process the given
files or directories.

Content added using \fBtracker index\fR will not be monitored for changes by
default, so you can end up with stale data in your search results.

Use \fB\-\-monitor\fR mode to avoid stale data. In this mode, the command will
not exit until CTRL+C is pressed, and on exit the content will be removed from
the store again.

It's often more convenient to configure the filesystem miner to index your
content automatically, rather than using this command.

.SH OPTIONS
.TP
.B \-\-no\-wait
Exit as soon as the files are queued for indexing. Errors during processing
will not be reported in this mode.
.TP
.B \-m, \-\-monitor
Trigger indexing and wait for CTRL+C, removing the data again on exit.

.SH SEE ALSO
.BR tracker (1).


CHANGELOG.md
Changelog
=========

2.2.99.1 - unreleased
----------------------------

  * The ~/Downloads folder is now indexed non-recursively by default.
      - This reduces the risk of Tracker unnecessarily crawling large
        directories of source code or video game data.
      - If users are storing data such as music or photos in subfolders of the
        Downloads folder, they will need to move it to an appropriate place
        (~/Music, ~/Pictures, etc.) so that it is still indexed.

**Translations:**

2.2.99.0 - 2019-08-07
----------------------------

  * Support for reading Musicbrainz metadata from audio files.
  * Tracker Writeback now uses GStreamer to write metadata to audio files,
    instead of depending on taglib directly.
  * Directories will now be ignored if they contain a file named `.nomedia`.
    A file named `.trackerignore` has the same effect, but the `.nomedia` file
    brings us in line with Android.
  * Removed obsolete 'max-media-art-width' setting.

2.2.2 - 2019-05-02
-------------------------

  * Multiple leak fixes
  * Functional tests now use python3
  * Fix text extractor handling of non-existent files
  * Fix indexing of tracks in FLAC files
  * Added new tests for file formats

**Translations:** hr

2.2.1 - 2019-03-06
-------------------------

  * Whitelist syscall fadvise64_64
  * Many fixes to functional tests
  * Fix failed functional tests being reported as successful

**Translations:** eo, fur, fr, it, ko, lv, nl, ru, sr

2.2.0 - 2019-02-20
-------------------------

  * Disable guarantee_metadata by default. It was the case on autotools
  * Stop merging translations to schema files
  * Test suite improvements
  * Meson build improvements

**Translations:** cs, da, de, es, gl, hu, id, lt, pl, pt_BR, ro, sl, sv, tr, zh_TW

  Highlighted changes since 2.1.x:
  --------------------------------

  * The functionality of tracker-miner-apps has been adopted by
    tracker-miner-fs/tracker-extract.
  * All usage of deprecated TrackerSparqlBuilder is gone.

2.2.0-alpha2 - 2019-02-06
--------------------------------
  * Fixes to desktop file indexing
  * Multiple build and compile warning fixes
  * Many test improvements and cleanups

**Translations:** fr

2.2.0-alpha1 - 2018-11-13
--------------------------------

  * The functionality of tracker-miner-apps has been adopted by
    tracker-miner-fs/tracker-extract.
  * Updated tracker-miner-fs and tracker-miner-rss to use TrackerResource
  * Support for building through autotools has been removed.
  * Plugged several leaks
  * Other many build and code cleanups and fixes

**Translations:** ru, sk, sr

2.1.4 - 2018-09-04
-------------------------

**Translations:** da, hr, lv

2.1.3 - 2018-08-30
-------------------------

  * Unbreak watch on domain ontology owner. Fixes miners spuriously exiting.

2.1.2 - 2018-08-30
-------------------------

  * Allow use of domain rules outside prefix
  * Add core-as-subproject as explicit meson option
  * Ensure utf8ness in TrackerResource helpers
  * Fix multiple cardinality issues in different extractors/properties
  * Other minor SPARQL correctness fixes

**Translations:** gl, id, ko

2.1.1 - 2018-08-15
-------------------------

  * tracker-miner-fs: Ignore mercurial repositories
  * build: Fix build with -Werror=format-security
  * tracker-extract: Ensure metadata strings are UTF-8 in JPEG/PNG extractors
  * tracker-extract: Do not add named destinations to PDF TOCs
  * tracker-extract: Ignore XMP image metadata on PDFs
  * tracker-extract: Fix ISO 8601 date string generation on gstreamer extractor

**Translations:** cs, el, fr, fur, hu, lt, sl, tr, zh_TW

2.1.0 - 2018-07-23
-------------------------

  * tests: Many updates and fixes to functional tests
  * tracker-extract: Better infrastructure for tests, based on JSON-LD
  * tracker-miner-fs: Restart tracker extract from the right domain
  * tracker-extract: Persistently avoid files that trigger recoverable
    errors, to avoid log spamming on restarts.

**Translations:** de, es, it, pl, pt_BR, ro, sv

2.0.5 - 2018-06-25
-------------------------

  * tracker-extract: Made daemon able to shutdown on inactivity, lifetime
    will be managed by tracker-miner-fs
  * tracker-stract: Fixes in generated SPARQL for geolocation data
  * build: Multiple improvements and library detection fixes

**Translations:** el, en_GB, hr, ro, ru

2.0.4 - 2018-02-07
-------------------------

  * build: Allow building tracker repo as a meson subproject
  * libtracker-common: Rename to libtracker-miners-common
  * libtracker-miners-common: Whitelist arm_fadvise64_64, getegid and getegid32 syscalls
  * tracker-extract: Add GExiv2-based extractor module for RAW files
  * tracker-extract: Blacklist gstreamer modules via plugin instead of via feature
  * tracker-extract: Blacklist video4linux2 gstreamer plugin
  * tracker-extract: Use enumerations for EXIF values
  * tracker-extract: Fix image pixel density conversions
  * tracker-miner-fs: Avoid setting rdf:types on empty files
  * meson: dependency check fixes

**Translations:** nb

2.0.3 - 2017-11-14
-------------------------

  * meson: Multiple cleanups and fixes
  * tracker-extract: Do not overwrite nie:dataSource

2.0.2 - 2017-10-05
-------------------------

  * tracker-miner-fs: Fix initialization when there's a startup timeout.
  * tracker-extract: Don't store 0-length strings on MSOffice metadata.

**Translations:** ca@valencia

2.0.1 - 2017-10-04
-------------------------

  * tracker-miner-apps: Ignore broken .desktop file symlinks
  * tracker-miner-fs: Ignore non-native mounts
  * tracker-extract: Fix insertion of white balance values in images
  * Compile warnings and Coverity issue fixes
  * Fix srcdir!=builddir builds

**Translations:** fr, sl

2.0.0 - 2017-09-12
--------------------------

  * Tracker shall from now on use semantic versioning
  * autotools/meson cleanups

  **Translations:** ca, da, de, eu, es, fur, hu, it, ko, lt, lv, nl, pt_BR, sk,

src/tracker/tracker-index.c
/*
 * Copyright (C) 2014, Lanedo <martyn@lanedo.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "config-miners.h"

#include <stdlib.h>
#include <errno.h>

#ifdef __sun
#include <procfs.h>
#endif

#include <glib.h>
#include <glib-unix.h>
#include <glib/gi18n.h>
#include <gio/gio.h>
#include <locale.h>

#include <libtracker-sparql/tracker-sparql.h>

#include "tracker-dbus.h"
#include "tracker-miner-manager.h"

static gboolean monitor_mode;
static gchar **filenames;

static GOptionEntry entries[] = {
	{ "monitor", 'm', 0, G_OPTION_ARG_NONE, &monitor_mode,
	  N_("Trigger indexing and wait for CTRL+C, removing the data again on exit."),
	  NULL },
	{ G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &filenames,
	  N_("FILE"),
	  N_("FILE") },
	{ NULL }
};

static gint
index_or_reindex_file (void)
{
	TrackerMinerManager *manager;
	GError *error = NULL;
	gchar **p;

	/* Auto-start the miners here if we need to */
	manager = tracker_miner_manager_new_full (TRUE, &error);
	if (!manager) {
		g_printerr (_("Could not (re)index file, manager could not be created, %s"),
		            error ? error->message : _("No error given"));
		g_printerr ("\n");
		g_clear_error (&error);
		return EXIT_FAILURE;
	}

	for (p = filenames; *p; p++) {
		GFile *file;

		file = g_file_new_for_commandline_arg (*p);
		tracker_miner_manager_index_file (manager, file, NULL, &error);

		if (error) {
			g_printerr ("%s: %s\n",
			            _("Could not (re)index file"),
			            error->message);
			g_error_free (error);
			return EXIT_FAILURE;
		}

		g_print ("%s\n", _("(Re)indexing file was successful"));
		g_object_unref (file);
	}

	g_object_unref (manager);

	return EXIT_SUCCESS;
}

static void
miner_file_processed (TrackerMinerManager *miner_manager,
                      const gchar         *miner,
                      const gchar         *uri,
                      const gboolean       status,
                      const gchar         *message,
                      GFile               *root)
{
	g_autoptr(GFile) file = NULL;
	g_autoptr(GError) error = NULL;
	g_autofree gchar *relative_path = NULL;

	file = g_file_new_for_uri (uri);

	relative_path = g_file_get_relative_path (root, file);

	if (g_file_equal (root, file) || relative_path != NULL) {
		/* File is a child of the root. */
		g_debug ("File processed: %s: %s, %i, %s\n", miner, uri, status, message);
	}
}

static gboolean
sigterm_cb (gpointer user_data)
{
	g_message ("Received signal");

	g_main_loop_quit (user_data);

	return G_SOURCE_REMOVE;
}

static gint
index_or_reindex_file_monitor_mode (void)
{
	g_autoptr(TrackerMinerManager) manager = NULL;
	g_autoptr(GMainLoop) main_loop = NULL;
	g_autoptr(GError) error = NULL;
	gchar **p;

	/* Check we were only passed directories. IndexFileForProcess doesn't work
	 * for files. */
	for (p = filenames; *p; p++) {
		g_autoptr(GFile) path = NULL;
		g_autoptr(GFileInfo) info = NULL;

		path = g_file_new_for_commandline_arg (*p);
		info = g_file_query_info (path,
		                          G_FILE_ATTRIBUTE_STANDARD_TYPE,
		                          G_FILE_QUERY_INFO_NONE,
		                          NULL,
		                          &error);

		if (error) {
			g_printerr ("%s: %s\n",
			            _("Error determining file type"),
			            error->message);
			return EXIT_FAILURE;
		}

		if (g_file_info_get_file_type (info) != G_FILE_TYPE_DIRECTORY) {
			g_printerr (_("Could not index file %s: in `--monitor` mode, "
			              "only directories can be indexed.\n"),
			            g_file_peek_path (path));
			return EXIT_FAILURE;
		}
	}

	/* Auto-start the miners here if we need to */
	manager = tracker_miner_manager_new_full (TRUE, &error);
	if (!manager) {
		g_printerr (_("Could not (re)index file, manager could not be created, %s"),
		            error ? error->message : _("No error given"));
		g_printerr ("\n");
		return EXIT_FAILURE;
	}

	for (p = filenames; *p; p++) {
		g_autoptr(GFile) dir = NULL;

		dir = g_file_new_for_commandline_arg (*p);

		g_signal_connect_data (G_OBJECT (manager),
		                       "miner-file-processed",
		                       G_CALLBACK (miner_file_processed),
		                       g_object_ref (dir),
		                       (GClosureNotify)g_object_unref,
		                       0);

		tracker_miner_manager_index_file_for_process (manager, dir, NULL, &error);

		if (error) {
			g_printerr ("%s: %s\n",
			            _("Could not (re)index directory"),
			            error->message);
			return EXIT_FAILURE;
		}

		g_print ("%s\n", _("(Re)indexing directory was successful"));
	}

	main_loop = g_main_loop_new (NULL, 0);

	g_unix_signal_add (SIGINT, sigterm_cb, main_loop);
	g_unix_signal_add (SIGTERM, sigterm_cb, main_loop);

	g_main_loop_run (main_loop);

	return EXIT_SUCCESS;
}

static int
index_run (void)
{
	if (monitor_mode) {
		return index_or_reindex_file_monitor_mode ();
	} else {
		return index_or_reindex_file ();
	}
}

int
main (int argc, const char **argv)
{
	GOptionContext *context;
	GError *error = NULL;
	const gchar *failed;

	setlocale (LC_ALL, "");

	bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	context = g_option_context_new (NULL);
	g_option_context_add_main_entries (context, entries, NULL);

	argv[0] = "tracker index";

	if (!g_option_context_parse (context, &argc, (char***) &argv, &error)) {
		g_printerr ("%s, %s\n", _("Unrecognized options"), error->message);
		g_error_free (error);
		g_option_context_free (context);
		return EXIT_FAILURE;
	}

	g_option_context_free (context);

	if (!filenames || g_strv_length (filenames) < 1) {
		failed = _("Please specify one or more locations to index.");
	} else {
		failed = NULL;
	}

	if (failed) {
		g_printerr ("%s\n", failed);
		return EXIT_FAILURE;
	}

	return index_run ();
}
